src/wait_group.cr

```diff
@@ -0,0 +1,92 @@
+require "fiber"
+require "crystal/spin_lock"
+require "crystal/pointer_linked_list"
+
+# Suspend execution until other fibers are done.
+#
+# This is a simpler and more efficient alternative to using a `Channel(Nil)`
+# then looping a number of times until we received N messages.
+#
+# Basic example:
+#
+# ```
+# require "wait_group"
+# wg = WaitGroup.new(5)
+#
+# 5.times do
+#   spawn do
+#     do_something
+#     wg.done # the fiber is done
+#   end
+# end
+#
+# # suspend the current fiber until the 5 fibers are done
+# wg.wait
```

Comment:
Somewhere, it would be nice to have some sort of text dealing with exceptions here. Or some description of who is expected to use this. Because if `do_something` raise an exception you get a stacktrace printed to stderr, but the program will neither finish nor crash. This is very much not a fantastic developer experience, especially not when run in an environment where the waitgroup itself is run in a fiber (like in a webserver).

Is this intended for end users to use directly or is it just a building block to build tools with that is less problematic? I know I will pick it up and use it for the latter, but some warning that it may create situations that are unfun to debug would be nice. This is a bigger problem than in go (which also have wait groups and expect end users to use them) because go don't have exceptions that will bobble up through the stack.

The good parts is that this code makes it a lot easier to build things on top of it that take care of this and doesn't have the sharp corners.

Reply:
Eventually, I would expect this to be mainly a building block for higher-level mechanics and users wouldn't deal with it directly unless implenting low-level concurrency stuff.

Reply:
You're right, I changed the example to wall `wg.done` inside an `ensure` block. It's a safer default example :)